### PR TITLE
Run All: Add support for code cells

### LINKED ISSUE #375: Cell boundaries
Hi there,

> Please note I've only used Hydrogen with IJavascript so I'm not sure if my observations only apply to this combination.

I think I understand that compared to a Jupyter notebook, it's difficult for Hydrogen to determine the boundaries of cells to enable the display of intermediary computation results in Atom when the whole notebook is ran? Is that the reason why intermediary results aren't displayed or updated?

For languages (like Javascript) where an optional line delimiter (`;`) can be used, wouldn't it be possible to use the presence of the delimiter as a cue to render the cell?

For other cases, or in addition to this, maybe Hydrogen could remember where existing cell runs have been requested and when a `run all` command is issued, it would update these bubbles (it seems that currently - with IJavascript - its only updating the final checkmark).

I guess that the more general case is to try to introspect the code to define the appropriate cell boundaries. This seems almost like something that would fit in a Kernel extension if there was interest upstream in allowing this type of freeform notebook. 

Let me know if that's unclear and I can explain more visually what I'm suggesting.

Cheers!

Jun

### PR DESCRIPTION
**Pull request by lgeiger**
Closes #375

### CODE REVIEW

**Comment by lgeiger** [comment]
Fixes #377

File: grammars/breakpoints.cson [hunk 1]
```diff
@@ -3,7 +3,7 @@
 'injectionSelector': 'comment.line'
 'patterns': [
   {
-    'match': '%%| %%| <codecell>| In\\[[0-9 ]+\\]:'
+    'match': '%%| %%| <codecell>| In\\[[0-9 ]*\\]:*'
```
Comment (n-riesco) [inline grammars/breakpoints.cson]: My bad, I should've written:
```cson
'match': '%%| %%| <codecell>| In\\[[0-9 ]*\\]:?'
```

File: lib/cell-manager.coffee [hunk 2]
```diff
@@ -59,7 +59,7 @@ module.exports = CellManager =
         escapedCommentStartString =
             escapeStringRegexp commentStartString.trimRight()
         regex = new RegExp(
-            escapedCommentStartString + '(%%| %%| <codecell>| In\[[0-9 ]+\]:)',
+            escapedCommentStartString + '(%%| %%| <codecell>| In\[[0-9 ]*\]:*)',
```
Comment (n-riesco) [inline lib/cell-manager.coffee]: ditto

File: grammars/breakpoints.cson (hunk 1 above)
Comment (lgeiger) [inline grammars/breakpoints.cson]: Oh yes thats better

**Comment by n-riesco** [comment]
We don't handle the case of inline comments  yet; e.g. : 
```coffeescript
a = 1 //%%
a += 1
//%%
```

But this is existing issue, so I'd leave the fix for another PR.

File: lib/cell-manager.coffee [hunk 3]
```diff
@@ -37,3 +27,39 @@ module.exports = CellManager =
...
+                breakpoints.push range.start.row
+
+        breakpoints.push editor.getLastBufferRow()
+
+        console.log 'CellManager: Breakpoints:', breakpoints
+
+        return breakpoints
+
+
+    getRegex: (editor, option = '') ->
```
Comment (n-riesco) [inline lib/cell-manager.coffee]: No need to act on this comment, but probably a `getRegexString` would've been a more versatile function.

File: lib/cell-manager.coffee [hunk 4]
```diff
@@ -37,3 +27,39 @@ module.exports = CellManager =
...
+
+    getBreakpoints: ->
+        editor = atom.workspace.getActiveTextEditor()
+        buffer = editor.getBuffer()
+        breakpoints = [0]
+
+        regex = @getRegex editor, 'g'
+        if regex?
+            buffer.scan regex, ({range}) ->
+                breakpoints.push range.start.row
```
Comment (n-riesco) [inline lib/cell-manager.coffee]: Again no need to act this comment, but to address the case of inline comments fully, here we may need to store `breakpoints.push range.start`/

Comment (lgeiger) [inline lib/cell-manager.coffee]: I'll address this in a future PR. But probably not today.

File: lib/cell-manager.coffee (hunk 3 above)
Comment (lgeiger) [inline lib/cell-manager.coffee]: You're right.  I wasn't really happy with the name either.

File: lib/cell-manager.coffee [hunk 5]
```diff
@@ -37,3 +27,39 @@ module.exports = CellManager =
...
+
+        breakpoints.push editor.getLastBufferRow()
+
+        console.log 'CellManager: Breakpoints:', breakpoints
+
+        return breakpoints
+
+
+    getRegexString: (editor, option = '') ->
```
Comment (n-riesco) [inline lib/cell-manager.coffee]: What I meant is that this function would be more versatile if it returned a string, and let the caller decide whether they want a Regex and what kind of options the regex should take.

Comment (n-riesco) [inline lib/cell-manager.coffee]: I'm sorry I didn't want to cause more unnecessary work for you.

Comment (lgeiger) [inline lib/cell-manager.coffee]: Oh now I understand.

No problem (thanks to `git rebase -i`) 😄

**Pull request merged**